### PR TITLE
Add context->scratchpad, for transient data

### PR DESCRIPTION
**Pull request by mmcclimon**
This came up in Topicbox, where we needed to keep track of a bunch of
updates during a single transaction, then do something at the end of it.

Each top-level transaction gets a new scratchpad, which is thrown away
when that transaction finishes.  This could maybe be an object, and have
better namespacing, but this works for now.

### CODE REVIEW

**Comment by mmcclimon** [comment]
What would maybe make more sense is to say `$context->scratchpad->for_call('setFoos')`, which would give you a hashref that's unique for that call.

I considered that, but because of the way the hook system works, a single call stack inside Ix (`ix_set`) turns into potentially a bunch of different call stacks inside the client (`ix_{creat,update,destroy}ed`, plus `ix_postprocess_{create,update,destroy}`), so if the scratchpad just returns a nonce reference, you still need somewhere in the client to store it, and it can't just be a lexical.

(This explanation might not make any sense).

File: lib/Ix/Context/WithAccount.pm [hunk 1]
```diff
@@ -61,6 +61,16 @@ sub _build_state ($self) {
   });
 }
 
+has scratchpad => (
```
Comment (wolfsage) [inline lib/Ix/Context/WithAccount.pm]: We should have some comments explaining what the scratchpad's scope is.

It looks like it's limited to a single JMAP Method and any JMAP calls made internally by that method, but not other Methods after the main method has finished executing

File: lib/Ix/Context/WithAccount.pm [hunk 2]
```diff
@@ -87,7 +97,12 @@ sub txn_do ($self, $code) {
     Carp::confess("We already have state before starting a transaction?!");
   }
 
-  # Start of a tree? Localize state so it goes away when we're done
+  # Start of a tree? Localize state and scratchpad so they go away when we're
```
Comment (wolfsage) [inline lib/Ix/Context/WithAccount.pm:110]: We're not actually localizing the scratchpad so this comment is wrong.

We could either fix the comment, or fix the code so we actually do localize the scratchpad (and then we don't have to clear it when done!)

**Review by wolfsage — COMMENTED** [review]
+1 with some more comments, and maybe a change

**Comment by mmcclimon** [comment]
Good call...both fixed now.

**Comment by wolfsage** [comment]
+1.

**Comment by mmcclimon** [comment]
We figured out a better way to do this in Topicbox, so I'm closing this now.

**Pull request closed without merging**